### PR TITLE
nixos/bind: add support for dynamic zone updates

### PR DESCRIPTION
**Pull request by Mewp**
###### Motivation for this change
Previously, if one created a zone with `{ file = pkgs.writeText ... }`, then set `allow-update { ... }` in extraConfig,
bind would be unhappy with the fact that the zones were in a read-only directory.

A broader motivation for this is that dynamic dns updates are useful for
ACME dns challenges, and it seems it's impossible to make a nixos configuration
that allows bind to receive zone updates from an ACME client.
At least not without supplying the initial zone file from outside nix.

###### Things done
This change copies the updateable zones to a writable location before
starting bind, so it can write zone changelogs along them.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](./CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
@grahamc I would really love if you could make 5 minutes to review this :heart:

File: nixos/modules/services/networking/bind.nix [hunk 1]
```diff
@@ -86,12 +92,19 @@ let
...
                 ${extraConfig}
               };
             '')
           (attrValues cfg.zones) }
     '';
 
+    copyUpdateableZones = optionalString ((length updateableZones) > 0) ''
+      mkdir -m 0755 -p /var/db/bind
+      chown ${bindUser} /var/db/bind
+      ${lib.concatMapStrings (zone: "[ -e /var/db/bind/${zone.name} ] && rm /var/db/bind/${zone.name}\n") updateableZones}
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/bind.nix:105]:
```suggestion
      ${lib.concatMapStrings (zone: "[ -e /var/db/bind/${zone.name} ]
      rm /var/db/bind/${zone.name}\n") updateableZones}
```

**Comment by RaitoBezarius** [comment]
@peti Could we get a review of this if possible?

File: nixos/modules/services/networking/bind.nix [hunk 2]
```diff
@@ -86,12 +92,19 @@ let
...
                 allow-query { any; };
+                ${optionalString allowUpdates "allow-update { key rndc-key; };"}
                 ${extraConfig}
               };
             '')
           (attrValues cfg.zones) }
     '';
 
+    copyUpdateableZones = optionalString ((length updateableZones) > 0) ''
+      mkdir -m 0755 -p /var/db/bind
```
Comment (peti) [inline nixos/modules/services/networking/bind.nix:103]: /var/db/bind feels inconsistent with /run/named. I'd prefer /var/cb/named here.

File: nixos/modules/services/networking/bind.nix (hunk 1 above)
Comment (peti) [inline nixos/modules/services/networking/bind.nix:105]: How about `rm -f /var/db/bind/${zone.name}`?

File: nixos/modules/services/networking/bind.nix [hunk 3]
```diff
@@ -86,12 +92,19 @@ let
...
                 }
                 allow-query { any; };
+                ${optionalString allowUpdates "allow-update { key rndc-key; };"}
                 ${extraConfig}
               };
             '')
           (attrValues cfg.zones) }
     '';
 
+    copyUpdateableZones = optionalString ((length updateableZones) > 0) ''
```
Comment (peti) [inline nixos/modules/services/networking/bind.nix:102]: When a zone is removed from the list of update-able zones, then its data file in /var/db/bind will not be removed either, which will result in dead files.

**Review by peti — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Closing due to being stale.

**Pull request closed without merging**